Allow for quotes in path for blob storage

adnanhemani opened the pull request:
<!--
    Possible security vulnerabilities: STOP here and contact security@apache.org instead!

    Please update the title of the PR with a meaningful message - do not leave it "empty" or "generated"
    Please update this summary field:

    The summary should cover these topics, if applicable:
    * the motivation for the change
    * a description of the status quo, for example the current behavior
    * the desired behavior
    * etc

    PR checklist:
    - Do a self-review of your code before opening a pull request
    - Make sure that there's good test coverage for the changes included in this PR
    - Run tests locally before pushing a PR (./gradlew check)
    - Code should have comments where applicable. Particularly hard-to-understand
      areas deserve good in-line documentation.
    - Include changes and enhancements to the documentation (in site/content/in-dev/unreleased)
    - For Work In Progress Pull Requests, please use the Draft PR feature.

    Make sure to add the information BELOW this comment.
    Everything in this comment will NOT be added to the PR description.
-->

Fixes #1545. Further context on that ticket.

Java URIs do not allow special characters, such as quotes, as part of [RFC 3986](https://datatracker.ietf.org/doc/html/rfc3986). As a result, it is not possible to use paths that contain such characters with Java URIs, which we use to compare storage locations when creating a new catalog.

This change refactors that code to only use URI if the location provided is a local filesystem path - else, it will use Hadoop's `Path` class in order to normalize the location.

adnanhemani commented:
Tagging the following here:
* @dimas-b since you opened the original issue (#1545)
* @eric-maynard since you authored a lot of the original code

Thanks!

snazy commented on polaris-core/src/main/java/org/apache/polaris/core/storage/StorageLocation.java:
```diff
@@ -19,8 +19,9 @@
 
 import jakarta.annotation.Nonnull;
-import java.net.URI;
+import org.apache.hadoop.yarn.logaggregation.LogCLIHelpers;
```
Yarn??

adnanhemani replied:
Yeah...no idea why IntelliJ did this to me 😅 thanks for catching!

snazy commented on polaris-core/src/main/java/org/apache/polaris/core/storage/StorageLocation.java:
```diff
@@ -19,8 +19,9 @@
-import java.net.URI;
+import org.apache.hadoop.yarn.logaggregation.LogCLIHelpers;
 import org.apache.polaris.core.storage.azure.AzureLocation;
+import org.apache.hadoop.fs.Path;
```
Not sure it's worth pulling in a lot of Hadoop dependencies. I prefer to keep Hadoop out of the code base.
There's an alternative though for this kind of issue [here in Nessie](https://github.com/projectnessie/nessie/blob/main/tools/storage/uri/src/main/java/org/projectnessie/storage/uri/StorageUri.java)

dimas-b replied:
+1 to avoid Hadoop dependencies

adnanhemani replied:
Thanks @snazy - I think the Nessie code is a bit more complex than what we require at this time, so I've created code inspired from the Nessie code you've linked. I hope this is reasonable - and thanks for the link!

snazy replied:
Not sure. Nessie's `StorageUri` was built for this issue, actually to work around the fact that Iceberg allows URIs that do not comply with RFC 2396. Why not use something that's already known to work?

dimas-b replied:
I'd also prefer reusing Nessie code in Polaris (we have other copy cases already). Nessie's `StorageUri` may seem to do more than what Polaris needs, but if / when we start dealing with more complex cases (e.g. Generic Table location), the extra functionality will probably be required.

dimas-b replied:
Side note: reviewing new code takes effort. Nessie's code already when through substantial reviews.

RussellSpitzer commented on polaris-core/src/main/java/org/apache/polaris/core/storage/StorageLocation.java:
```diff
@@ -19,8 +19,9 @@
+import org.apache.hadoop.fs.Path;
 
 /** An abstraction over a storage location */
 public class StorageLocation {
```
note, we need tests for the new case being supported

adnanhemani replied:
Yes, good call out. Included in the latest commit!

eric-maynard commented:
+1 to the the other comments that we should avoid a Hadoop dependency in polaris-core if possible, although you can find [some past comments](https://github.com/apache/polaris/pull/344#discussion_r1788128038) where I suggested we may eventually want to use something like it.

If we ever do cut over to something like `Path`, we need to be exhaustive about it, removing everywhere we treat the path as a string and making all the relevant methods work with `Path`.

For now, the solution we use for Azure is this [AzureLocation](https://github.com/apache/polaris/blame/9d663b2b9cd74a82ef50ed160879082bc11e9a05/polaris-core/src/main/java/org/apache/polaris/core/storage/azure/AzureLocation.java#L27) -- I wonder if we can try to use something similar here. Certainly, we should try to avoid a mix of `String`, `Path`, and `AzureLocation`.

adnanhemani commented:
I originally used Hadoop `Path` class because it is the closest thing to what we want and we already bring the Hadoop Client API dependency in other modules, such as Service. I don't personally think adding a client API to this module is the worst thing in the world, but not adding it just makes us to have to normalize the paths ourselves.

So I'm not strong on doing this either way but given there's unanimous pushback on it, I've implemented the normalization manually in the latest push.

@eric-maynard - I thought about creating new `XYZLocation` types, but I found them to just end up being very simplistic classes after drafting the changes. S3 and GCS don't have the same level of complexity that Azure has, so I've implemented the normalization in `StorageLocation` itself. Please let me know if this is something you feel strongly about. Thanks all!

eric-maynard commented on polaris-core/src/main/java/org/apache/polaris/core/storage/StorageLocation.java:
```diff
@@ -104,4 +112,22 @@ public boolean isChildOf(StorageLocation potentialParent) {
     }
   }
+
+  private static String scheme(String location) {
```
nit: method name

adnanhemani replied:
Changed to `getScheme`

eric-maynard replied:
Did this get missed?

adnanhemani replied:
Sorry - seems like things didn't push :( Will push with the next set of comments.

eric-maynard commented on polaris-core/src/main/java/org/apache/polaris/core/storage/StorageLocation.java:
```diff
@@ -41,16 +44,21 @@ public static StorageLocation of(String location) {
+    String path;
+    if (scheme == null) {
+      path = location;
+      scheme = "file";
```
We don't have an enum or something for this?

adnanhemani replied:
Unfortunately not, we are parsing them from the user-provided input. I tried hard-typing these and rejecting any schemes that are not explicitly supported (but including file formats that don't explicitly state a scheme) and then a barrage of test cases started failing due to weird input we've put in - some of them intentionally, others not. So I'm not really in favor of making such a massive shift, it will dramatically increase the scope of this change.

eric-maynard commented on polaris-core/src/main/java/org/apache/polaris/core/storage/StorageLocation.java:
```diff
@@ -41,16 +44,21 @@ public static StorageLocation of(String location) {
+      path = location.substring(scheme.length() + 1);
+    }
+    path = normalizePath(path);
+    if (scheme.equals("file")) {
```
ditto, let's at least put this in a const

adnanhemani replied:
Roger that, that's reasonable.

eric-maynard commented on polaris-core/src/main/java/org/apache/polaris/core/storage/StorageLocation.java:
```diff
@@ -41,16 +44,21 @@ public static StorageLocation of(String location) {
+    if (scheme.equals("file")) {
+      this.location = URI.create(LOCAL_PATH_PREFIX + path).toString();
+    } else {
+      this.location = scheme + "://" + path;
```
Doesn't this duplicate the `://`?

adnanhemani replied:
No - the scheme will be without the "://", as we parse right up to the ":" for it. The path is normalized in the `normalizePath` function such that we will remove any starting slashes and the parsing for path starts after the ":" character.

eric-maynard commented on polaris-core/src/main/java/org/apache/polaris/core/storage/StorageLocation.java:
```diff
@@ -41,16 +44,21 @@ public static StorageLocation of(String location) {
+      path = location;
+      scheme = "file";
     } else {
-      this.location = new Path(location).toString();
```
Looks like this branch used to cover relative paths, but now it won't

adnanhemani replied:
That's correct. Trying to deal with these is only going to muddy the waters further here if we try to implement it manually :( And I don't think this is a case that we really need to care heavily for cloud storage providers given that they are really just pseudo-filesystems.

If this is a strong requirement, I heavily recommend that we go back to Hadoop's `Path` class instead of trying this manually....

dimas-b commented on polaris-core/src/test/java/org/apache/polaris/service/storage/StorageLocationTest.java:
```diff
@@ -19,17 +19,41 @@
 
 import org.apache.polaris.core.storage.StorageLocation;
-import org.assertj.core.api.Assertions;
+import org.junit.jupiter.api.Assertions;
```
Why not `assertj`? They are generally more user-friendly, IMHO

adnanhemani replied:
Earlier, I was testing some assertionThrows and JUnit was much easier to work with for those imo. But now that those are gone, I agree for these cases - I'll change them back.

dimas-b commented on polaris-core/src/test/java/org/apache/polaris/service/storage/StorageLocationTest.java:
```diff
@@ -19,17 +19,41 @@
+    Assertions.assertEquals(testStorageLocation, baseStorageLocation);
+
+    testStorageLocation = StorageLocation.of("s3://////path'/////*to===///\"file");
+    Assertions.assertEquals("s3://path'/*to===/\"file", testStorageLocation.toString());
```
nit: in this URI `path'` is not a "path" by "authority" per [RFC 3986](https://www.rfc-editor.org/rfc/rfc3986)

dimas-b commented on polaris-core/src/test/java/org/apache/polaris/service/storage/StorageLocationTest.java:
```diff
@@ -19,17 +19,41 @@
+  public void testGcpLocations() {
+    StorageLocation baseStorageLocation = StorageLocation.of("gcs://path/to/file");
+    StorageLocation testStorageLocation = StorageLocation.of("gcs://path/to/////file");
+    Assertions.assertEquals(testStorageLocation, baseStorageLocation);
```
Is this test materially different from `testAwsLocations`?

adnanhemani replied:
In spirit, no. Let me merge this case and the AWS cases.

dimas-b commented on polaris-core/src/test/java/org/apache/polaris/service/storage/StorageLocationTest.java:
```diff
@@ -19,17 +19,41 @@
+  public void testAwsLocations() {
+    StorageLocation baseStorageLocation = StorageLocation.of("s3://path/to/file");
+    StorageLocation testStorageLocation = StorageLocation.of("s3://path/to/////file");
+    Assertions.assertEquals(testStorageLocation, baseStorageLocation);
```
This test is good, but I believe more test cases are needed to validate the handling of upper / lower case letters and leading / trailing slashes and the special cases of missing scheme.

adnanhemani replied:
Good call out - let me add them

dimas-b commented:
Re: #1545 : this PR fixes part of the problem, but I'm not sure it fixes all that needs to be fixed. I believe the use case in #1545 needs to be tested end-to-end (with reading from tables) to be sure it is fixed.

adnanhemani commented on polaris-core/src/main/java/org/apache/polaris/core/storage/StorageLocation.java:
```diff
@@ -104,4 +112,22 @@ public boolean isChildOf(StorageLocation potentialParent) {
+  private static String scheme(String location) {
+    int schemePos = location.indexOf(':');
+    if (schemePos > 0) {
+      return location.substring(0, schemePos).toLowerCase(Locale.ROOT);
```
The lowercasing is from the Nessie inspired normalization. I don't think it will make it necessarily incorrect - but seems like something we do want.

For example, the following files _should_ get flagged as having a parent-child relationship:
file:///a/b/c/d
FILE:///a/b/c/d

adnanhemani commented:
> I believe the use case in #1545 needs to be tested end-to-end (with reading from tables) to be sure it is fixed.

@dimas-b - I think that's a reasonable ask. Are you saying that we should do this manually once or that this needs an integration test?

dimas-b commented:
I mostly meant that this PR should not claim that it `Fixes #1545` (in the descriptio) :)

I did not mean to expand the scope of this PR. We can merge this incremental improvement, re-test the issue manually and follow-up with more PRs and integration tests as required.

eric-maynard commented on polaris-core/src/main/java/org/apache/polaris/core/storage/StorageLocation.java:
```diff
@@ -99,4 +111,23 @@ public boolean isChildOf(StorageLocation potentialParent) {
+    if (schemePos > 0) {
+      return location.substring(0, schemePos).toLowerCase(Locale.ROOT);
+    }
+    return null;
```
Should we throw here?

adnanhemani replied:
No - this is passed early in the making of the `StorageLocation` and can, as a result, get filesystem locations like `/var/sample/location`. This is how we're figuring out that this is a local file path.

eric-maynard commented on polaris-core/src/main/java/org/apache/polaris/core/storage/StorageLocation.java:
```diff
@@ -99,4 +111,23 @@ public boolean isChildOf(StorageLocation potentialParent) {
+  }
+
+  protected String normalizePath(String path) {
+    path = SLASHES.matcher(path).replaceAll("/");
```
I don't think this is always correct. IIRC on Azure `/foo/bar` is not the same as `/foo//bar`

eric-maynard replied:
It looks like this is just a special case for `file:///...` which should really just be treated as an empty authority

adnanhemani replied:
Ahh yes, great catch. Can't believe I forgot about this but you're correct,  S3 is the same as well. I will make changes to make sure this is only happening for local files then.

eric-maynard commented on polaris-core/src/main/java/org/apache/polaris/core/storage/StorageLocation.java:
```diff
@@ -99,4 +111,23 @@ public boolean isChildOf(StorageLocation potentialParent) {
+
+  protected String normalizePath(String path) {
+    path = SLASHES.matcher(path).replaceAll("/");
+    if (path.endsWith("/") || path.startsWith("/")) {
+      path =
+          path.substring(
+              path.startsWith("/") && path.length() > 1 ? 1 : 0,
+              path.endsWith("/") ? path.length() - 1 : path.length());
+    }
```
`StringUtils.strip`

eric-maynard commented on polaris-core/src/test/java/org/apache/polaris/service/storage/StorageLocationTest.java:
```diff
@@ -19,17 +19,41 @@
+    testStorageLocation = StorageLocation.of("gcs://////path/////to///file");
+    Assertions.assertEquals(testStorageLocation, baseStorageLocation);
+
+    testStorageLocation = StorageLocation.of("gcs://////path'/////*to===///\"file");
+    Assertions.assertEquals("gcs://path'/*to===/\"file", testStorageLocation.toString());
```
How does it work right now in Polaris?

adnanhemani replied:
It's not correct. Actually when looking through Java URI code - we are not currently applying _any_ normalization to URLs...

And I'm not sure that normalizing URLs is entirely necessary either. We've worked through the comments above that double slashes shouldn't be modified and similarly, "." and ".." also shouldn't be normalized for blob storage. 

As a result, I think this might be trying to do too much. I'm going to stop work on this and push a new branch that simply just takes the URI create off for the Blob Storage locations and adds additional tests. Will link that PR here when I push it.

adnanhemani commented on polaris-core/src/main/java/org/apache/polaris/core/storage/StorageLocation.java:
```diff
@@ -99,4 +111,23 @@ public boolean isChildOf(StorageLocation potentialParent) {
+    return null;
+  }
+
+  protected String normalizePath(String path) {
```
I had never heard of this term before, but this is correct! I will change it, thank you!

adnanhemani replied:
Seems like both terms are synonyms - at least I'm not seeing a difference between these two terms. Can you please explain the difference or provide a link to something that does? For reference, this is what [Google](https://www.google.com/search?q=canonicalize+vs+normalize&client=safari&sca_esv=861d6c36a75495fa&rls=en&sxsrf=AHTn8zrDV0AEV8cAvsCBTIMeqtVihnblHw%3A1747499926443&ei=lrsoaMPtGvDokPIPoK2qqAU&ved=0ahUKEwjDkPHC-KqNAxVwNEQIHaCWClUQ4dUDCBA&oq=canonicalize+vs+normalize&gs_lp=Egxnd3Mtd2l6LXNlcnAiGWNhbm9uaWNhbGl6ZSB2cyBub3JtYWxpemUyChAAGLADGNYEGEcyChAAGLADGNYEGEcyChAAGLADGNYEGEcyChAAGLADGNYEGEcyChAAGLADGNYEGEcyChAAGLADGNYEGEcyChAAGLADGNYEGEcyChAAGLADGNYEGEdI2gdQAFgAcAF4AZABAJgBAKABAKoBALgBDMgBAJgCAaACCpgDAIgGAZAGCJIHATGgBwCyBwC4BwA&sclient=gws-wiz-serp) is pulling up on this.

eric-maynard replied:
[Canonicalization](https://hadoop.apache.org/docs/stable/api/org/apache/hadoop/fs/FileSystem.html#canonicalizeUri-java.net.URI-) usually means adding all the missing parts, such as the scheme and authority, whereas [normalization](https://stackoverflow.com/a/10823859) / resolution usually just means making a path absolute.

adnanhemani commented:
Created #1604. Will keep this PR open until #1604 merges or at least gets confirmed as the right way to go - but please do not review this PR until further notice.

eric-maynard commented:
Closing in favor of #1604
